mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,360 @@
+use mm2_net::transport::GuiAuthValidationGenerator;
+use std::collections::HashMap;
+use std::sync::atomic::AtomicBool;
+use std::sync::{atomic::{AtomicUsize, Ordering},
+                Arc};
+use tokio_tungstenite_wasm::WebSocketStream;
+use web3::error::{Error, TransportError};
+use web3::helpers::to_string;
+use web3::{helpers::build_request, RequestId, Transport};
+
+const REQUEST_TIMEOUT_AS_SEC: u64 = 10;
+
+#[derive(Clone, Debug)]
+pub(crate) struct WebsocketTransportNode {
+    pub(crate) uri: http::Uri,
+    pub(crate) gui_auth: bool,
+}
+
+#[derive(Clone, Debug)]
+pub struct WebsocketTransport {
+    request_id: Arc<AtomicUsize>,
+    pub(crate) last_request_failed: Arc<AtomicBool>,
+    node: WebsocketTransportNode,
+    event_handlers: Vec<RpcTransportEventHandlerShared>,
+    pub(crate) gui_auth_validation_generator: Option<GuiAuthValidationGenerator>,
+    responses: Arc<SafeMapPtr>,
+    controller_channel: Arc<ControllerChannel>,
+    connection_guard: Arc<AsyncMutex<()>>,
+}
+
+#[derive(Debug)]
+struct ControllerChannel {
+    tx: Arc<AsyncMutex<UnboundedSender<ControllerMessage>>>,
+    rx: Arc<AsyncMutex<UnboundedReceiver<ControllerMessage>>>,
+}
+
+enum ControllerMessage {
+    Request(WsRequest),
+    Close,
+}
+
+#[derive(Debug)]
+struct WsRequest {
+    serialized_request: String,
+    request_id: RequestId,
+    response_notifier: oneshot::Sender<()>,
+}
+
+/// A wrapper type for raw pointers used as a mutable Send & Sync HashMap reference.
+///
+/// Safety notes:
+///
+/// The implemented algorithm for socket request-response is already thread-safe,
+/// so we don't care about race conditions.
+///
+/// As for deallocations, see the `Drop` implementation below.
+#[derive(Debug)]
+struct SafeMapPtr(*mut HashMap<RequestId, Vec<u8>>);
+
+impl Drop for SafeMapPtr {
+    fn drop(&mut self) {
+        // Let the compiler do the job.
+        let _ = unsafe { Box::from_raw(self.0) };
+    }
+}
+
+unsafe impl Send for SafeMapPtr {}
+unsafe impl Sync for SafeMapPtr {}
+
+impl WebsocketTransport {
+    pub(crate) fn with_event_handlers(
+        node: WebsocketTransportNode,
+        event_handlers: Vec<RpcTransportEventHandlerShared>,
+    ) -> Self {
+        let (req_tx, req_rx) = futures::channel::mpsc::unbounded();
+
+        WebsocketTransport {
+            node,
+            event_handlers,
+            responses: Arc::new(SafeMapPtr(Box::into_raw(Default::default()))),
+            request_id: Arc::new(AtomicUsize::new(1)),
+            controller_channel: ControllerChannel {
+                tx: Arc::new(AsyncMutex::new(req_tx)),
+                rx: Arc::new(AsyncMutex::new(req_rx)),
+            }
+            .into(),
+            connection_guard: Arc::new(AsyncMutex::new(())),
+            gui_auth_validation_generator: None,
+            last_request_failed: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    pub(crate) async fn start_connection_loop(self, expires_at: Option<Instant>) {
```

Comment:
I suggest to refactor `start_connection_loop` a bit to improve readability.

I think you can move `attempt_to_establish_socket_connection` function to `impl WebsocketTransport` from inner `attempt_to_establish_socket_connection` functionality. 
Also it worth moving logic related to: 
```rust
// KEEPALIVE HANDLING
_ = keepalive_interval.next().fuse() =>

// SEND REQUESTS  
request = req_rx.next().fuse() =>

// HANDLE RESPONSES SENT FROM USER  
message = wsocket.next().fuse() =>
```
into new private functions.

So you will have this.
```rust
impl WebsocketTransport {
    async fn attempt_to_establish_socket_connection(  
    &self,  
    address: String,  
    max_attempts: u32,  
    mut sleep_duration_on_failure: f64,  
) -> tokio_tungstenite_wasm::Result<WebSocketStream> {  
    const MAX_SLEEP_DURATION: f64 = 32.0;  
    let mut attempts = 0;  
  
    loop {  
        match tokio_tungstenite_wasm::connect(address.clone()).await {  
            Ok(ws) => return Ok(ws),  
            Err(e) => {  
                attempts += 1;  
                if attempts > max_attempts {  
                    return Err(e);  
                }  
  
                Timer::sleep(sleep_duration_on_failure).await;  
                sleep_duration_on_failure = (sleep_duration_on_failure * 2.0).min(MAX_SLEEP_DURATION);  
            },  
        };  
    }  
}  
  
pub(crate) async fn start_connection_loop(self, expires_at: Option<Instant>) {  
    let _guard = self.connection_guard.lock().await;  
  
    const MAX_ATTEMPTS: u32 = 3;  
    const SLEEP_DURATION: f64 = 1.;  
    const SIMPLE_REQUEST: &str = r#"{"jsonrpc":"2.0","method":"net_version","params":[],"id": 0 }"#;  
    const KEEPALIVE_DURATION: Duration = Duration::from_secs(10);  
  
    let mut response_notifiers: ExpirableMap<RequestId, oneshot::Sender<()>> = ExpirableMap::default();  
    let mut wsocket = match self  
        .attempt_to_establish_socket_connection(self.node.uri.to_string(), MAX_ATTEMPTS, SLEEP_DURATION)  
        .await  
    {  
        Ok(ws) => ws,  
        Err(e) => {  
            log::error!("Connection could not established for {}. Error {e}", self.node.uri);  
            return;  
        },  
    };  
  
    let mut keepalive_interval = Ticker::new(KEEPALIVE_DURATION);  
    let mut req_rx = self.controller_channel.rx.lock().await;  
  
    loop {  
        futures_util::select! {  
            _ = keepalive_interval.next().fuse() => {  
                self.handle_keepalive(&mut wsocket, &mut response_notifiers, expires_at, SIMPLE_REQUEST, MAX_ATTEMPTS, SLEEP_DURATION).await;  
            },  
            request = req_rx.next().fuse() => {  
                self.handle_incoming_request(&mut wsocket, &mut response_notifiers, request).await;  
            },  
            message = wsocket.next().fuse() => {  
                self.handle_websocket_message(&mut response_notifiers, message).await;  
            },  
        }  
    }  
}  
  
async fn handle_keepalive(  
    &self,  
    _wsocket: &mut WebSocketStream,  
    _response_notifiers: &mut ExpirableMap<RequestId, oneshot::Sender<()>>,  
    _expires_at: Option<Instant>,  
    _simple_request: &str,  
    _max_attempts: u32,  
    _sleep_duration: f64,  
) {  
    todo!()  
}  
  
async fn handle_incoming_request(  
    &self,  
    _wsocket: &mut WebSocketStream,  
    _response_notifiers: &mut ExpirableMap<RequestId, oneshot::Sender<()>>,  
    _request: Option<ControllerMessage>,  
) {  
    todo!()  
}  
  
async fn handle_websocket_message(  
    &self,  
    _response_notifiers: &mut ExpirableMap<RequestId, oneshot::Sender<()>>,  
    _message: Option<Result<tokio_tungstenite_wasm::Message, tokio_tungstenite_wasm::Error>>,  
) {  
    todo!()  
}
}
```

Well its fine, to leave `attempt_to_establish_socket_connection` as an inner function of general `fn start_connection_loop`. 
My main suggestion is to extract the logic for `futures_util::select!` into separate functions. 

I think this is easier to read
```rust
loop {  
        futures_util::select! {  
            _ = keepalive_interval.next().fuse() => {  
                self.handle_keepalive(&mut wsocket, &mut response_notifiers, expires_at, SIMPLE_REQUEST, MAX_ATTEMPTS, SLEEP_DURATION).await;  
            },  
            request = req_rx.next().fuse() => {  
                self.handle_incoming_request(&mut wsocket, &mut response_notifiers, request).await;  
            },  
            message = wsocket.next().fuse() => {  
                self.handle_websocket_message(&mut response_notifiers, message).await;  
            },  
        }  
    } 
```

Reply:
As I explained at https://github.com/KomodoPlatform/komodo-defi-framework/pull/2058#discussion_r1489434126, break/continue based loop flows are not easy to seperated. I implemented something called `OuterAction` to achieve seperating functions from the loop. I hope it looks good.

Reply:
> It's not possible to parse select! flows into that many functions as they based on loops and continues.

Oh, I see. Thanks for the refactoring :pray: Looks good!